syntax = "proto3";

service KademliaService {
	rpc Ping(PingRequest) returns (PingResponse);
	rpc FindNode(FindNodeRequest) returns (FindNodeResponse);
	rpc Store(StoreRequest) returns (StoreResponse);
	rpc Get(GetRequest) returns (GetResponse);
}

message ConnectionInfo{
	string id = 1;
	string ip = 2;
	uint32 port = 3;
}

message PingRequest{
	uint64 magic = 1;
	ConnectionInfo caller = 2;
}

message PingResponse{
	uint64 magic = 1;
}

message FindNodeRequest{
	ConnectionInfo caller = 1;
	string destination = 2;
}

message FindNodeResponse{
	repeated ConnectionInfo closest = 1;
}

message StoreRequest{
	uint64 magic = 1;
	ConnectionInfo caller = 2;
	string key = 3;
	bytes value = 4;
}

message StoreResponse{
	uint64 magic = 1;
}

message GetRequest{
	uint64 magic = 1;
	ConnectionInfo caller = 2;
	string key = 3;
}

message GetResponse{
	uint64 magic = 1;
	bytes value = 2;
}
